README.md
<!-- markdownlint-configure-file {
  "MD013": {
    "code_blocks": false,
    "tables": false
  },
  "MD033": false,
  "MD041": false
} -->
<div align="center">

# NeatRepeat
### for Logic Pro Scripter

*A note so nice, I played it twice.*
<br/>
*With passion. And precision.*


[![PRs welcome](https://img.shields.io/badge/PRs-welcome-ff69b4.svg?style=flat-square)](https://github.com/AdamAdamsMusic/NeatRepeat/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22)
[![Downloads][downloads-badge]][releases]

---

![Demo by Vince Webb](NeatRepeat.gif)

</div>

## Features
- Choose a re-trigger key from the menu, or click **LEARN** to allow script to… well, LEARN, **which key you want to use** for it.
- Play a note, then press the re-trigger key to **play the same note again instantly**.
- **Preserve the same velocity** as the original note, or use the re-trigger key properties to **add more nuance to your performance**.
- Choose whether you want the re-triggered notes to be **played on the same MIDI channel** or rather another one, so that **they can overlap**! (*instrument-dependent*)
- Have fun and get creative with your performances – works especially well on **percussive instruments/one-shots**! (*experimentation encouraged*)

## Getting started
- Download the [latest release][latest]
- Move the .pst file to ~/Music/Audio Music Apps/Plug-In Settings/Scripter
- Add "Scripter" in the MIDI FX plugin section in Logic
- (Optional, but strongly suggested) Click on the MIDI FX slot just below "Scripter" and enable "Record MIDI to Track Here"
- Load up the preset from the list

## See it in action

[![Play fast repeated notes with ease (NeatRepeat) – Vince Webb](http://i3.ytimg.com/vi/kl1RwhATTzA/hqdefault.jpg)](http://www.youtube.com/watch?v=kl1RwhATTzA "Play fast repeated notes with ease (NeatRepeat)")

## About

One day, [Vince Webb](https://www.vincewebb.com/) was looking for a tech helper to realize his idea – he wanted to have a script that he could use just like the "re-trigger" functions in Orange Tree Samples libraries, but for **EVERY** instrument.

That is how *NeatRepeat* started. A script that allows for re-triggering played notes, so that the performances on the MIDI keyboard can be more realistic and – let's admit it – more fun!

MIDI keyboard has a size constraint. It's hard to quickly press one key in rapid succession using both hands. But sometimes, what is needed in the performance is exactly that. This is where this script comes it – allowing you to specify which key you'd want to use for re-triggering the key and letting you comfortably using two hands to ~~HAMMER IN~~ record your performance.

## Known issues
- MIDI protocol specification, more specifically the lack of "note instance". MIDI only allows sending NoteOn/NoteOff messages, without actual note instances. That is understandable – you have one key for every note. This functionality – having many instances of the exact same note on the exact same MIDI channel – was not really envisioned, because… *why would you need it*?! So when lots of identical messages are sent in quick succession, some may get cut off or hang indefinitely.

## Contributing

First off, thanks for taking the time to contribute! Contributions are what makes the open-source community such an amazing place to learn, inspire, and create. Any contributions you make will benefit everybody else and are **greatly appreciated**.

Please try to create bug reports that are:

- _Reproducible._ Include steps to reproduce the problem.
- _Specific._ Include as much detail as possible: which version, what environment, etc.
- _Unique._ Do not duplicate existing opened issues.
- _Scoped to a Single Bug._ One bug per report.

## Support

As much as I like doing this and taking part in an open-source community, I have to underline that **I try to provide support and improvements in my spare time**. You can reach out to me at one of the following places:

- [GitHub discussions](https://github.com/AdamAdamsMusic/NeatRepeat/discussions)
- The email which is located [in my GitHub profile](https://github.com/AdamAdamsMusic)

## License

This project is licensed under the **GNU General Public License v3**.

See [LICENSE](LICENSE) for more information.

# Ackowledgements

Thanks to [Vince Webb](https://vincewebb.com) and the great [TEAMMATES community](https://www.facebook.com/groups/assistthecomposer) for making this happen!

<div align=center>
<img src="https://yt3.googleusercontent.com/gcS_1zgM3cKwHdEJdEvcPWLsBA6rsnnCaKBJscq9T9O8kt5I35mkT9mA3W-6osjxsOk9cNUczaY=s0" style="width: 300px; height: 300px; border-radius: 150px">
</div>


[downloads-badge]: https://img.shields.io/github/downloads/AdamAdamsMusic/NeatRepeat/total?logo=github&logoColor=white&style=flat-square
[latest]: https://github.com/AdamAdamsMusic/NeatRepeat/releases/latest
[releases]: https://github.com/AdamAdamsMusic/NeatRepeat/releases


CHANGELOG
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog],
and this project adheres to [Semantic Versioning].

## [Unreleased]

## [1.0.0] - 2023-11-07

### Added

- Added changelog tracking

### Changed

- Changed the name to NeatRepeat

## 0.1.5 - 2023-11-07

### Added

- Different (MIDI) Ch. for re-trigger
  - Since not having separate note instances is a limitation of the MIDI protocol, when the script sends a lot of MIDI NoteOn/NoteOff messages for the exact same key – it gets really confusing and some notes are cut off, while the other hang indefinitely. The workaround is to send the re-triggered notes on a different MIDI channel, so they don't interfere. This allows for proper re-triggering without artifacts, but comes with its own limitations (especially visible with VEP/multitimbral setup users).

## 0.1.4 - 2023-11-07

### Fixed

- Script now allows any other MIDI message (e.g. MIDI CC) to be bypassed without processing, instead of filtering them out

## 0.1.3 - 2023-11-07

### Added

- Script now kills (sends NoteOff for) the previously sounding note of the same number, in order to trigger another one – removing the overlay issue

## 0.1.2 - 2023-11-07

### Changed

- Now the trigger key note behaves as it should, instead of having a NoteOff sent after a fixed amount of time

## 0.1.1 - 2023-11-07

### Added

- Delay between NoteOn and NoteOff is now calculated automatically, based on the note you hit before the trigger key

### Removed

- Removed the NoteOff delay slider

### Fixed

- Choosing trigger key from the menu didn't work

## 0.1.0 - 2023-11-05

- initial release

<!-- Links -->
[keep a changelog]: https://keepachangelog.com/en/1.0.0/
[semantic versioning]: https://semver.org/spec/v2.0.0.html

<!-- Versions -->
[unreleased]: https://github.com/AdamAdamsMusic/NeatRepeat/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/AdamAdamsMusic/NeatRepeat/releases/tag/v1.0.0

NeatRepeat.js
///////////////////////////////////////////////////
//                                               //
//                   NeatRepeat                  //
//          Logic Pro X Scripter script          //
//                                               //
//             Made for: Vince Webb              //
//          https://www.vincewebb.com/           //
//                                               //
//  Author: Adam Adams (adam@adam-adams.com)     //
//  Version: v1.0.0 (2023-11-07)                 //
//                                               //
///////////////////////////////////////////////////

var midiNotes = [
  "C-2",
  "C#-2",
  "D-2",
  "D#-2",
  "E-2",
  "F-2",
  "F#-2",
  "G-2",
  "G#-2",
  "A-2",
  "A#-2",
  "B-2",
  "C-1",
  "C#-1",
  "D-1",
  "D#-1",
  "E-1",
  "F-1",
  "F#-1",
  "G-1",
  "G#-1",
  "A-1",
  "A#-1",
  "B-1",
  "C0",
  "C#0",
  "D0",
  "D#0",
  "E0",
  "F0",
  "F#0",
  "G0",
  "G#0",
  "A0",
  "A#0",
  "B0",
  "C1",
  "C#1",
  "D1",
  "D#1",
  "E1",
  "F1",
  "F#1",
  "G1",
  "G#1",
  "A1",
  "A#1",
  "B1",
  "C2",
  "C#2",
  "D2",
  "D#2",
  "E2",
  "F2",
  "F#2",
  "G2",
  "G#2",
  "A2",
  "A#2",
  "B2",
  "C3",
  "C#3",
  "D3",
  "D#3",
  "E3",
  "F3",
  "F#3",
  "G3",
  "G#3",
  "A3",
  "A#3",
  "B3",
  "C4",
  "C#4",
  "D4",
  "D#4",
  "E4",
  "F4",
  "F#4",
  "G4",
  "G#4",
  "A4",
  "A#4",
  "B4",
  "C5",
  "C#5",
  "D5",
  "D#5",
  "E5",
  "F5",
  "F#5",
  "G5",
  "G#5",
  "A5",
  "A#5",
  "B5",
  "C6",
  "C#6",
  "D6",
  "D#6",
  "E6",
  "F6",
  "F#6",
  "G6",
  "G#6",
  "A6",
  "A#6",
  "B6",
  "C7",
  "C#7",
  "D7",
  "D#7",
  "E7",
  "F7",
  "F#7",
  "G7",
  "G#7",
  "A7",
  "A#7",
  "B7",
  "C8",
  "C#8",
  "D8",
  "D#8",
  "E8",
  "F8",
  "F#8",
  "G8",
];

var PluginParameters = [
  {
    name: "Trigger Note",
    type: "menu",
    valueStrings: midiNotes,
    defaultValue: 48,
  },
  { name: "Learn", type: "momentary", defaultValue: 0 },
  { name: "Use original velocity", type: "checkbox", defaultValue: 0 },
  { name: "Different Ch. for retriggers", type: "checkbox", defaultValue: 0 },
];

var lastNote;
var channelOffset = 1;
var multiChannel = false;
var triggerNote = GetParameter("Trigger Note");
var learning = false;

function createEvent(type, pitch, velocity, channel) {
  var event = new type();
  event.pitch = MIDI.normalizeData(pitch);
  event.velocity = MIDI.normalizeData(velocity);
  event.channel = MIDI.normalizeData(channel);
  return event;
}

function sendEvent(event) {
  event.send();
  event.trace();
}

function HandleMIDI(event) {
  if (event.channel == 16) {
    -channelOffset;
  }
  if (event instanceof NoteOn || event instanceof NoteOff) {
    var currentTriggerNote = GetParameter("Trigger Note");
    if (triggerNote !== currentTriggerNote) {
      triggerNote = currentTriggerNote;
      SetParameter("Trigger Note", triggerNote);
    }
    if (event instanceof NoteOn) {
      if (learning) {
        triggerNote = event.pitch;
        learning = false;
        SetParameter("Trigger Note", triggerNote);
      } else if (event.pitch == triggerNote && lastNote != null) {
        var newEvent = createEvent(NoteOff, lastNote.pitch, 0, lastNote.channel);
        sendEvent(newEvent);
        newEvent = createEvent(NoteOn, lastNote.pitch, GetParameter("Use original velocity") ? lastNote.velocity : event.velocity, GetParameter("Different Ch. for retriggers") ? lastNote.channel + channelOffset : lastNote.channel);
        sendEvent(newEvent);
      } else {
        lastNote = event;
        sendEvent(event);
      }
    } else if (event instanceof NoteOff) {
      if (event.pitch == triggerNote) {
        var newEvent = createEvent(NoteOff, lastNote.pitch, 0, GetParameter("Different Ch. for retriggers") ? lastNote.channel + channelOffset : lastNote.channel);
        sendEvent(newEvent);
      } else {
        sendEvent(event);
      }
    }
  } else {
    sendEvent(event);
  }
}

function ParameterChanged(param, value) {
  if (param == 1) {
    learning = true;
  }
}


logic_retrigger_last_note_v1.00.js
var lastNote = null;
var triggerNote = null;
var learning = false;

var midiNotes = [
    "C-2", "C#-2", "D-2", "D#-2", "E-2", "F-2", "F#-2", "G-2", "G#-2", "A-2", "A#-2", "B-2",
    "C-1", "C#-1", "D-1", "D#-1", "E-1", "F-1", "F#-1", "G-1", "G#-1", "A-1", "A#-1", "B-1",
    "C0", "C#0", "D0", "D#0", "E0", "F0", "F#0", "G0", "G#0", "A0", "A#0", "B0",
    "C1", "C#1", "D1", "D#1", "E1", "F1", "F#1", "G1", "G#1", "A1", "A#1", "B1",
    "C2", "C#2", "D2", "D#2", "E2", "F2", "F#2", "G2", "G#2", "A2", "A#2", "B2",
    "C3", "C#3", "D3", "D#3", "E3", "F3", "F#3", "G3", "G#3", "A3", "A#3", "B3",
    "C4", "C#4", "D4", "D#4", "E4", "F4", "F#4", "G4", "G#4", "A4", "A#4", "B4",
    "C5", "C#5", "D5", "D#5", "E5", "F5", "F#5", "G5", "G#5", "A5", "A#5", "B5",
    "C6", "C#6", "D6", "D#6", "E6", "F6", "F#6", "G6", "G#6", "A6", "A#6", "B6",
    "C7", "C#7", "D7", "D#7", "E7", "F7", "F#7", "G7", "G#7", "A7", "A#7", "B7",
    "C8", "C#8", "D8", "D#8", "E8", "F8", "F#8", "G8"
];

var PluginParameters = [
    {name:"Trigger Note", type:"menu", valueStrings:midiNotes, defaultValue:48},
    {name:"Learn", type:"momentary", defaultValue:0},
    {name:"NoteOff delay", type:"lin", numberOfSteps:398, minValue:10, maxValue:2000, defaultValue:100}
    {name:"Use original velocity", type:"check", }
];

function HandleMIDI(event) {
    if (event instanceof NoteOn) {
        if (learning) {
            triggerNote = event.pitch;
            learning = false;
            SetParameter("Trigger Note", triggerNote);
        } else if (event.pitch == triggerNote) {
            if (lastNote != null) {
                // Preparing new event
                var newEvent = new NoteOn();
                // Setting up all the properties as the lastNote
                newEvent.pitch = MIDI.normalizeData(lastNote.pitch);
                newEvent.velocity = MIDI.normalizeData(lastNote.velocity);
                newEvent.channel = MIDI.normalizeData(lastNote.channel);
                // Send it
                newEvent.send();
                newEvent.trace();
                // Have to send noteOff too...
                newEvent = new NoteOff();
                newEvent.pitch = MIDI.normalizeData(lastNote.pitch);
                newEvent.velocity = MIDI.normalizeData(lastNote.velocity);
                newEvent.channel = MIDI.normalizeData(lastNote.channel);
                // We'll give it customizable delay
                newEvent.sendAfterMilliseconds(GetParameter("NoteOff delay"));
                newEvent.trace();
            }
        } else {
            // Store the last played note
            lastNote = event;
            event.send();
            event.trace();
        }
    } else if (event instanceof NoteOff && lastNote != null && event.pitch == lastNote.pitch) {
        // NoteOff event for the last played note, update lastNote to null
        event.send();
        event.trace();
    } else {
        event.send();
    }
}

function ParameterChanged(param, value) {
    if (param == 1) {
        learning = true;
    }
}



logic_retrigger_last_note.js
///////////////////////////////////////////////////
//                                               //
//           logic_retrigger_last_note           //
//          Logic Pro X Scripter script          //
//                                               //
//             Made for: Vince Webb              //
//          https://www.vincewebb.com/           //
//                                               //
//  Author: Adam Adams (adam@adam-adams.com)     //
//  Version: v1.01 (2023-11-07)                  //
//                                               //
///////////////////////////////////////////////////

var midiNotes = [
  "C-2",
  "C#-2",
  "D-2",
  "D#-2",
  "E-2",
  "F-2",
  "F#-2",
  "G-2",
  "G#-2",
  "A-2",
  "A#-2",
  "B-2",
  "C-1",
  "C#-1",
  "D-1",
  "D#-1",
  "E-1",
  "F-1",
  "F#-1",
  "G-1",
  "G#-1",
  "A-1",
  "A#-1",
  "B-1",
  "C0",
  "C#0",
  "D0",
  "D#0",
  "E0",
  "F0",
  "F#0",
  "G0",
  "G#0",
  "A0",
  "A#0",
  "B0",
  "C1",
  "C#1",
  "D1",
  "D#1",
  "E1",
  "F1",
  "F#1",
  "G1",
  "G#1",
  "A1",
  "A#1",
  "B1",
  "C2",
  "C#2",
  "D2",
  "D#2",
  "E2",
  "F2",
  "F#2",
  "G2",
  "G#2",
  "A2",
  "A#2",
  "B2",
  "C3",
  "C#3",
  "D3",
  "D#3",
  "E3",
  "F3",
  "F#3",
  "G3",
  "G#3",
  "A3",
  "A#3",
  "B3",
  "C4",
  "C#4",
  "D4",
  "D#4",
  "E4",
  "F4",
  "F#4",
  "G4",
  "G#4",
  "A4",
  "A#4",
  "B4",
  "C5",
  "C#5",
  "D5",
  "D#5",
  "E5",
  "F5",
  "F#5",
  "G5",
  "G#5",
  "A5",
  "A#5",
  "B5",
  "C6",
  "C#6",
  "D6",
  "D#6",
  "E6",
  "F6",
  "F#6",
  "G6",
  "G#6",
  "A6",
  "A#6",
  "B6",
  "C7",
  "C#7",
  "D7",
  "D#7",
  "E7",
  "F7",
  "F#7",
  "G7",
  "G#7",
  "A7",
  "A#7",
  "B7",
  "C8",
  "C#8",
  "D8",
  "D#8",
  "E8",
  "F8",
  "F#8",
  "G8",
];

var PluginParameters = [
  {
    name: "Trigger Note",
    type: "menu",
    valueStrings: midiNotes,
    defaultValue: 48,
  },
  { name: "Learn", type: "momentary", defaultValue: 0 },
  {
    name: "NoteOff delay",
    type: "lin",
    numberOfSteps: 384,
    minValue: 80,
    maxValue: 2000,
    defaultValue: 100,
  },
  { name: "Use original velocity", type: "checkbox", defaultValue: 0 },
];

var lastNote;
var noteOnTime;
var delay;
var triggerNote = GetParameter("Trigger Note");
var learning = false;

function HandleMIDI(event) {
  var currentTriggerNote = GetParameter("Trigger Note");
  if (triggerNote === null || triggerNote !== currentTriggerNote) {
    triggerNote = currentTriggerNote;
    SetParameter("Trigger Note", triggerNote);
  }
  if (event instanceof NoteOn) {
    noteOnTime = new Date().getTime();
    if (learning) {
      triggerNote = event.pitch;
      learning = false;
      SetParameter("Trigger Note", triggerNote);
    } else if (event.pitch == triggerNote) {
      if (lastNote != null) {
        // Preparing new event
        var newEvent = new NoteOn();
        // Setting up all the properties as the lastNote
        newEvent.pitch = MIDI.normalizeData(lastNote.pitch);
        if (GetParameter("Use original velocity")) {
          newEvent.velocity = MIDI.normalizeData(lastNote.velocity);
        } else {
          newEvent.velocity = MIDI.normalizeData(event.velocity);
        }
        newEvent.channel = MIDI.normalizeData(lastNote.channel);
        // Send it
        newEvent.send();
        newEvent.trace();
        // Have to send noteOff too...
        newEvent = new NoteOff();
        newEvent.pitch = MIDI.normalizeData(lastNote.pitch);
        newEvent.velocity = MIDI.normalizeData(0);
        newEvent.channel = MIDI.normalizeData(lastNote.channel);
        // We'll give it customizable delay
        Trace("Measured delay: " + delay);
        newEvent.sendAfterMilliseconds(delay);
        newEvent.trace();
      }
    } else {
      // Store the last played note
      lastNote = event;
      event.send();
      event.trace();
    }
  } else if (
    event instanceof NoteOff &&
    lastNote != null &&
    event.pitch == lastNote.pitch
  ) {
    // NoteOff event for the last played note, update lastNote to null
    delay = new Date().getTime() - noteOnTime;
    event.send();
    event.trace();
  } else {
    event.send();
  }
}

function ParameterChanged(param, value) {
  if (param == 1) {
    learning = true;
  }
}
